### PR TITLE
Feat/new metrics

### PR DESCRIPTION
**Pull request by thib-s**
added new metrics:
- ProvableAvgRobustness: return the average size of provable radius ( for both binary and multiclass ) there is a parameter to choose if misclassified examples gives a 0 radius or a negative radius.
- ProvableRobustAccuracy: return the accuracy that can be guaranteed for a given epsilon.

Also added tests for these metrics

### CODE REVIEW

File: tests/test_metrics.py [hunk 1]
```diff
@@ -66,8 +64,8 @@ def test_provable_vs_adjusted(self):
         x1 = tf.random.normal((n, 10), mean=0.0, stddev=0.1)
         x2 = tf.random.normal((n, 10), mean=1.0, stddev=0.1)
         for lip_cst in [1.0, 0.5, 2.0]:
-            bar = BinaryAdjustedRobustness(lip_cst)
-            bpr = BinaryProvableRobustness(lip_cst)
+            bpr = ProvableAvgRobustness(lip_cst, disjoint, negative_robustness=False)
+            bar = ProvableAvgRobustness(lip_cst, disjoint, negative_robustness=True)
```
Comment (cofri) [inline tests/test_metrics.py:80]: I found names `bpr` and `bar` confusing because they are not binary metrics anymore...

**Review by cofri — CHANGES_REQUESTED** [review]

File: tests/test_metrics.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+)
+import os
+
+
+def check_serialization(nb_class, loss):
+    m = Sequential([Input(10), Dense(nb_class)])
+    m.compile(optimizer=SGD(), loss=loss)
+    name = loss.__class__.__name__
+    path = os.path.join("logs", "losses", name)
+    m.save(path)
```
[the remark this answers is not in the document]
Comment (thib-s) [reply tests/test_metrics.py:27]: that's true, this should be done in test_layers, test_condense to be consistent
( it is preferable if these other files are edited in a separate PR )

File: tests/test_metrics.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                self.assertNotAlmostEqual(
+                    l1,
+                    l2,
+                    4,
+                    msg="provable and adjusted robustness must give different values"
+                    " when y_true!=y_pred",
+                )
+
+        x1 = tf.random.normal((n, 10), mean=0.0, stddev=0.1)
+        x2 = tf.random.normal((n, 10), mean=1.0, stddev=0.1)
```
[the remark this answers is not in the document]
Comment (thib-s) [reply tests/test_metrics.py]: the tests have been updated to check if each metric handle both the 1-D format `(batch_size, )` and the 2-D format `(batch_size, 1)`

File: tests/test_metrics.py (hunk 1 above)
Comment (thib-s) [reply tests/test_metrics.py:80]: the metrics are now handling both binary and multiclass now :-)

File: deel/lip/metrics.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        else:
+            delta_fct = _delta_binary
+            self.certificate_factor = self.lip_const
+        return tf.reduce_mean(
+            tf.cast(
+                (delta_fct(y_true, y_pred) / self.certificate_factor) > self.epsilon,
+                tf.float32,
+            )
+        )
+
```
Comment (cofri) [inline deel/lip/metrics.py:109]: In the same fashion as the element-wise losses, the `call()` should return an element-wise loss (here 1 or 0 if robustess larger than epsilon) and the `Reduction.AUTO` will compute the average. It is possible to remove the `reduce_mean()` here.

Moreover, I think it's better to cast to `y_pred.dtype`, instead of hardcoded tf.float32

File: deel/lip/metrics.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+            self.certificate_factor = math.sqrt(2) * lip_const
+        super(ProvableRobustAccuracy, self).__init__(reduction, name)
+
+    def call(self, y_true, y_pred):
+        shape = y_true.shape
+        if len(shape) == 2 and (shape[-1] > 1):
+            delta_fct = _delta_multiclass
+        else:
+            delta_fct = _delta_binary
+            self.certificate_factor = self.lip_const
```
Comment (cofri) [inline deel/lip/metrics.py:102]: That's sound odd to me to set the `self.certificate_factor` here in the `call()`. I don't know what is the simplest and cleanest option... Maybe an argument for binary vs. multiclass setup? Or two separate metrics, such as in losses.py?

**Review by cofri — CHANGES_REQUESTED** [review]

File: deel/lip/metrics.py (hunk 5 above)
Comment (thib-s) [reply deel/lip/metrics.py:102]: yeah, in combination with a `@tf.function`, `y_true.shape` is considered as a constant and thus the built graph don't embed the `if`, and there is no overhead. This wouldn't be the case if we used a `tf.shape(y_true)` (which is an operator).

**Review by cofri — APPROVED** [review]
